draft-robert-mls-extensions.md

```diff
@@ -93,6 +102,173 @@ to recognize which messages contain AppAcks.  The application can also have
+with the sender and the specific recipients – which might not be possible or
+desired. Instead, targeted messages define the format and encryption of a
+message that is sent from a member of an existing group to another member of
+that group.
+
+The goal is to provide a one-shot messaging mechanism that provides
+confidentiality and authentication.
+
+### Format
+
+This extensions extens the MLS protocol to include a new message type,
+`TargetedMessage` in `WireFormat` and `MLSMessage`:
+
+~~~ tls
+enum {
+  ...
+  mls_targeted_message(6),
+  ...
+  (255)
+} WireFormat;
+
+struct {
+    ProtocolVersion version = mls10;
+    WireFormat wire_format;
+    select (MLSMessage.wire_format) {
+        ...
+        case mls_targeted_message:
+            TargetedMessage targeted_message;
+    }
+} MLSMessage;
+~~~
+
+The `TargetedMessage` message type is defined as follows:
+
+~~~ tls
+struct {
+  opaque group_id<V>;
+  uint64 epoch;
+  uint32 recipient_leaf_index;
+  opaque authenticated_data<V>;
+  opaque encrypted_sender_data<V>;
+  opaque encrypted_targeted_message_content<V>;
+} TargetedMessage;
+
+enum {
+  HPKEAuth,
+  Signature,
+} TargetedMessageAuthScheme;
+
+struct {
+  TargetedMessageAuthScheme authentication_scheme;
+  select (authentication_scheme) {
+    case HPKEAuth:
+    case Signature:
+      opaque signature<V>;
+  }
+  HPKECiphertext ciphertext;
+} TargetedMessageContent;
+
+struct {
+  opaque group_id<V>;
+  uint64 epoch;
+  uint32 recipient_leaf_index;
+  opaque authenticated_data<V>;
+  opaque encrypted_sender_data<V>;
+  TargetedMessageAuthScheme authentication_scheme;
+} TargetedMessageTBM;
+
+struct {
+  TargetedMessageTBM targeted_message_tbm;
+  HPKECiphertext hpke_ciphertext;
+} TargetedMessageTBS;
+
+struct {
+  opaque group_id<V>;
+  uint64 epoch;
+  opaque label<V> = "MLS 1.0 targeted message psk";
+} PSKId;
+~~~
+
+### Encryption
+
+Targeted messages use HPKE to encrypt the message content between two leaves.
+The HPKE keys of the `LeafNode` are used to that effect, namely the
+`encryption_key` field.
+
+In addition, the sender data encryption from section 7.3.2 {{mls-protocol}} is
+used to encrypt `MLSSenderData`. `MLSSenderData.leaf_index` is the leaf index of
+the sender. The `MLSSenderData.generation` field is not used and MUST be set to
+0.
+
+### Authentication
+
+For ciphersuites that support it, HPKE `mode_auth_psk` is used for
+authentication. For other ciphersuites, HPKE `mode_psk` is used along with a
+signature. The authentication scheme is indicated by the `authentication_scheme`
+field in `TargetedMessageContent`. See {{guidance-on-authentication-schemes}}
+for more information.
+
+For the PSK part of the authentication, clients export a dedicated secret:
+
+```
+targeted_message_psk = MLS-Exporter("targeted message psk", "", KDF.Nh)
+```
+
+#### Authentication with HPKE
+
+The sender MUST set the authentication scheme to
+`TargetedMessageAuthScheme.HPKEAuth`.
```

Comment:
Using `HPKEAuth` when it is actually `HPKEAuthPsk` is a little confusing.

Reply:
I see how that's confusing. I left out the PSK part because it is common to both authentication modes and therefore not a differentiator. I'll try to think of a better name.